File: MEPmodel_bio.m
%---------12 free parameters (with RC population)------------
% p(1:5): membrane resistances R of MNs
% p(6:7): connectivity W_exc (MNs->RC)
% p(8:9): connectivity R.*W_inh (RC->MNs)
% p(10) : RC.th (threshold of RC sigmoid function)
% p(11) : Tmu (refractory period of muscle unit)
% p(12) : AMPA weight (for DI-wave input)
%
%---------7 free parameters (without RC population)--------
% p(1:5): membrane resistances R of MNs
% p(6)  : Tmu (refractory period of muscle unit)
% p(7)  : AMPA weight (for DI-wave input)

function [simMEP,ref] = MEPmodel_bio(p,ref,plotOn)

    if nargin<3
        plotOn=0;
    end
   
    ref = update_model_bio(ref,p);
    sim = MEPmodel_bio_core(ref.model); % simulation
    ref = cal_error(ref,sim);
    simMEP = ref.sim.simMEP2;
    
    if plotOn        

        % show fitted parameters in the search boundary
        plot_param_panel(p,ref); 

        % summarize goodness-of-fit of MEPs across TMS intensities
        plot_summary(p,ref); 

        % select a TMS intensity to show MN-RC interaction
        idx=length(ref.intensities); %1; 
        plot_TMScondition(ref,idx); 
    end
end
%==========================================================================
function ref = update_model_bio(ref,p)
    switch ref.model.withRC
        case 0
            R     = p(1:5);
            E1    = 0; 
            E2    = 0; 
            I1    = 0; 
            I2    = 0;
            RCth  = 10;
            Tmu   = p(6);
            AMPAw = p(7);
        case 1
            R     = p(1:5);
            E1    = p(6); 
            E2    = p(7); 
            I1    = p(8); 
            I2    = p(9);
            RCth  = p(10);
            Tmu   = p(11);
            AMPAw = p(12);
    end
    %----- membrane resistances of MNs -----
    ref.model.R = gen_resistance_mono(R);
    %----- Wexc (MNs -> RC) -----
    ref.model.Wexc = linspace(E1,E2,100);   
    %----- R*Winh (RC->MNs)-----
    ref.model.RWinh = linspace(I1,I2,100); 
    %----- RC sigmoid ------
    ref.model.rc = RC_setting(RCth);
    %----- MUAP-----
    ref.model.Tmu       = Tmu;  % ms, refractory period 
    %----- DI-waves -----
    ref.model.AMPAweight= AMPAw;
    ref.model.DIwaveConv= ref.model.DIwaveConv_AMPA* ref.model.AMPAweight + ...
                          ref.model.DIwaveConv_NMDA* (1 - ref.model.AMPAweight);

end
%==========================================================================
% Reshaw cell population params
function rc=RC_setting(p)
   
    % RC sigmoid function
    rc.r     = 10;
    rc.v_thr = p;
    rc.fmax  = 1;

    % synaptic kernel (RC->MN, Glycine receptor)
    taur=1; % ms rise
    tauf=6; % ms fall
    rc.tau1=taur*tauf/(taur+tauf); 
    rc.tau2=tauf;
    rc.h=(rc.tau2-rc.tau1)/rc.tau1/rc.tau2; 

end

%==========================================================================
function ref = cal_error(ref,sim)

    y0=ref.y0;
    t0=ref.t0;
    simMEP=sim.simMEP'; % [500 x 10]
    t=sim.t;
    %---------------------
    % align peaks of simMEP and MEP
    [~,i1]=max(y0(:,end)); % largest TMS intensity
    [~,i2]=max(simMEP(:,end));
    axonalDelay=max(0,t0(i1)-t(i2)); % >=0 ms 
    simMEP = interp1(t + axonalDelay,simMEP,t0,'linear',0); % shift d
    %------------------------------------
    if max(simMEP(:))>0 % avoid devide by zero
        simMEP2=simMEP/max(simMEP(:))*max(y0(:));
    else
        simMEP2=simMEP;
    end
    error=simMEP2(:)-y0(:);
    NRMSD=norm(y0(:)-simMEP2(:))/sqrt(length(t0))/(max(y0(:))-min(y0(:)));
    R2=1-sumsqr(y0(:)-simMEP2(:))/sumsqr(y0(:)-mean(y0(:)));
    %---------------------
    ref.NRMSD=NRMSD;
    sim.simMEP2=simMEP2;
    ref.model.axonalDelay=axonalDelay;
    ref.R2=R2;
    ref.error=error;
    ref.sim=sim;
end

File: MEPmodel_bio.py
import numpy as np
from scipy.interpolate import interp1d
from plot_summary import plot_summary
from plot_param_panel import plot_param_panel
from plot_TMScondition import plot_TMScondition
from MEPmodel_bio_core import MEPmodel_bio_core
from gen_resistance_mono import gen_resistance_mono

# --------- Main function ---------
def MEPmodel_bio(p, ref, plotOn=0):

    ref = update_model_bio(ref, p)
    sim = MEPmodel_bio_core(ref["model"])  # simulation 
    ref = cal_error(ref, sim)
    simMEP = ref["sim"]["simMEP2"]

    if plotOn:
        plot_param_panel(p, ref)
        plot_summary(p, ref)

        idx = len(ref["intensities"])  # same as MATLAB
        plot_TMScondition(ref, idx)

    return simMEP, ref


# ==========================================================================
def update_model_bio(ref, p):

    if ref["model"]["withRC"] == 0:
        R     = p[0:5]
        E1    = 0
        E2    = 0
        I1    = 0
        I2    = 0
        RCth  = 10
        Tmu   = p[5]
        AMPAw = p[6]

    elif ref["model"]["withRC"] == 1:
        R     = p[0:5]
        E1    = p[5]
        E2    = p[6]
        I1    = p[7]
        I2    = p[8]
        RCth  = p[9]
        Tmu   = p[10]
        AMPAw = p[11]

    # ----- membrane resistances of MNs -----
    ref["model"]["R"] = gen_resistance_mono(R)

    # ----- Wexc (MNs -> RC) -----
    ref["model"]["Wexc"] = np.linspace(E1, E2, 100)

    # ----- R*Winh (RC->MNs) -----
    ref["model"]["RWinh"] = np.linspace(I1, I2, 100)

    # ----- RC sigmoid -----
    ref["model"]["rc"] = RC_setting(RCth)

    # ----- MUAP -----
    ref["model"]["Tmu"] = Tmu

    # ----- DI-waves -----
    ref["model"]["AMPAweight"] = AMPAw
    ref["model"]["DIwaveConv"] = (
        ref["model"]["DIwaveConv_AMPA"] * ref["model"]["AMPAweight"]
        + ref["model"]["DIwaveConv_NMDA"] * (1 - ref["model"]["AMPAweight"])
    )

    # ----- Motor neuron -----
    ref["model"]["tauLIF"] = 10
    ref["model"]["Eexc"] = 0
    ref["model"]["Einh"] = -75
    ref["model"]["V_rest"] = -65
    ref["model"]["V_thr"] = -55
    ref["model"]["T_ref"] = 2
    ref["model"]["fastAChRweight"] = 0.5

    return ref


# ==========================================================================
# Renshaw cell population params
def RC_setting(p):

    rc = {}

    # RC sigmoid function
    rc["r"] = 10
    rc["v_thr"] = p
    rc["fmax"] = 1

    # synaptic kernel (RC->MN, Glycine receptor)
    taur = 1
    tauf = 6
    rc["tau1"] = taur * tauf / (taur + tauf)
    rc["tau2"] = tauf
    rc["h"] = (rc["tau2"] - rc["tau1"]) / rc["tau1"] / rc["tau2"]

    return rc


# ==========================================================================
def cal_error(ref, sim):

    y0 = ref["y0"]
    t0 = ref["t0"]

    simMEP = sim["simMEP"].T  # transpose like MATLAB '
    t = sim["t"]

    # align peaks
    i1 = np.argmax(y0[:, -1])
    i2 = np.argmax(simMEP[:, -1])

    axonalDelay = max(0, t0[i1] - t[i2])

    # interpolation (MATLAB interp1 equivalent)
    f_interp = interp1d(
        t + axonalDelay,
        simMEP,
        axis=0,
        kind="linear",
        bounds_error=False,
        fill_value=0
    )
    simMEP = f_interp(t0)

    # normalization
    if np.max(simMEP) > 0:
        simMEP2 = simMEP / np.max(simMEP) * np.max(y0)
    else:
        simMEP2 = simMEP

    error = simMEP2 - y0

    NRMSD = (
        np.linalg.norm(y0 - simMEP2)
        / np.sqrt(len(t0))
        / (np.max(y0) - np.min(y0))
    )

    R2 = 1 - np.sum((y0 - simMEP2) ** 2) / np.sum((y0 - np.mean(y0)) ** 2)

    ref["NRMSD"] = NRMSD
    sim["simMEP2"] = simMEP2
    ref["model"]["axonalDelay"] = axonalDelay
    ref["R2"] = R2
    ref["error"] = error
    ref["sim"] = sim

    return ref